wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/jpa/JpaFactoryDao.java

```diff
@@ -107,42 +111,89 @@ public FactoryImpl getById(String id) throws NotFoundException, ServerException
   }
 
   @Override
-  @Transactional
-  public List<FactoryImpl> getByAttribute(
+  @Transactional(rollbackOn = {RuntimeException.class, ServerException.class})
+  public Page<FactoryImpl> getByAttributes(
       int maxItems, int skipCount, List<Pair<String, String>> attributes) throws ServerException {
+    checkArgument(maxItems >= 0, "The number of items to return can't be negative.");
+    checkArgument(
+        skipCount >= 0 && skipCount <= Integer.MAX_VALUE,
+        "The number of items to skip can't be negative or greater than " + Integer.MAX_VALUE);
     try {
       LOG.debug(
           "FactoryDao#getByAttributes #maxItems: {} #skipCount: {}, #attributes: {}",
           maxItems,
           skipCount,
           attributes);
-      final Map<String, String> params = new HashMap<>();
-      String query = "SELECT factory FROM Factory factory";
-      if (!attributes.isEmpty()) {
-        final StringJoiner matcher = new StringJoiner(" AND ", " WHERE ", " ");
-        int i = 0;
-        for (Pair<String, String> attribute : attributes) {
-          final String parameterName = "parameterName" + i++;
-          params.put(parameterName, attribute.second);
-          matcher.add("factory." + attribute.first + " = :" + parameterName);
-        }
-        query = query + matcher;
+      final long count = countFactoriesByAttributes(attributes);
+      if (count == 0) {
+        return new Page<>(emptyList(), skipCount, maxItems, count);
       }
-      final TypedQuery<FactoryImpl> typedQuery =
-          managerProvider
-              .get()
-              .createQuery(query, FactoryImpl.class)
-              .setFirstResult(skipCount)
-              .setMaxResults(maxItems);
-      for (Map.Entry<String, String> entry : params.entrySet()) {
-        typedQuery.setParameter(entry.getKey(), entry.getValue());
-      }
-      return typedQuery.getResultList().stream().map(FactoryImpl::new).collect(Collectors.toList());
+      List<FactoryImpl> result = getFactoriesByAttributes(maxItems, skipCount, attributes);
+      return new Page<>(result, skipCount, maxItems, count);
     } catch (RuntimeException ex) {
       throw new ServerException(ex.getLocalizedMessage(), ex);
     }
   }
 
+  @Override
+  @Transactional(rollbackOn = {RuntimeException.class, ServerException.class})
+  public Page<FactoryImpl> getByUser(String userId, int maxItems, long skipCount) {
+    requireNonNull(userId);
+    final Pair<String, String> factoryCreator = Pair.of("creator.userId", userId);
+    long totalCount = countFactoriesByAttributes(singletonList(factoryCreator));
```

Comment:
As far as I see runtime exception may be thrown by an entity manager inside of `countFactoriesByAttributes` method. It must be mapped to `ServerException` but it IS NOT. Am I right?